qiskit/providers/ibmq/ibmqbackend.py

```diff
@@ -746,6 +751,97 @@ def __repr__(self) -> str:
+            self,
+            circuits: Union[QasmQobj, PulseQobj, QuantumCircuit, Schedule,
+                            List[Union[QuantumCircuit, Schedule]]]
+    ) -> None:
+        """Raise a DeprecationWarning if any circuit contains an 'id' instruction.
+
+        Additionally, if 'delay' is a 'supported_instruction', replace each 'id'
+        instruction (in-place) with the equivalent ('sx'-length) 'delay' instruction.
+
+        Args:
+            circuits: The individual or list of :class:`~qiskit.circuits.QuantumCircuit` or
+                :class:`~qiskit.pulse.Schedule` objects passed to
+                :meth:`IBMQBackend.run()<IBMQBackend.run>`. Modified in-place.
+
+        Returns:
+            None
+        """
+
+        if isinstance(circuits, PulseQobj):
+            return
+
+        id_support = 'id' in getattr(self.configuration(), 'basis_gates', [])
+        delay_support = 'delay' in getattr(self.configuration(), 'supported_instructions', [])
+
+        if isinstance(circuits, QasmQobj):
+            circuit_has_id = any(instr.name == 'id'
+                                 for experiment in circuits.experiments
+                                 for instr in experiment.instructions)
+        else:
+            if not isinstance(circuits, List):
+                circuits = [circuits]
+
+            circuit_has_id = any(instr.name == 'id'
+                                 for circuit in circuits
+                                 if isinstance(circuit, QuantumCircuit)
+                                 for instr, qargs, cargs in circuit.data)
+
+        if not circuit_has_id:
+            return
+
+        if not self.id_warning_issued:
+            if id_support:
+                warnings.warn("Support for the 'id' instruction has been deprecated "
+                              "from IBM hardware backends. Any 'id' instructions "
+                              "will be replaced with their equivalent 'delay' instruction. "
+                              "Please use the 'delay' instruction instead.", DeprecationWarning,
+                              stacklevel=4)
+            elif delay_support:
+                warnings.warn("Support for the 'id' instruction has been removed "
+                              "from IBM hardware backends. Any 'id' instructions "
+                              "will be replaced with their equivalent 'delay' instruction. "
+                              "Please use the 'delay' instruction instead.", DeprecationWarning,
+                              stacklevel=4)
+            else:
+                warnings.warn("Support for the 'id' instruction has been removed "
+                              "from IBM hardware backends. Please use the 'delay' "
+                              "instruction instead.", DeprecationWarning,
+                              stacklevel=4)
+
+            self.id_warning_issued = True
+
+        if not delay_support:
+            return
```

Comment:
The warnings on line 797 and 803 say `id` is being replaced by `delay`, but that's not true if `delay` is not reported as supported by the backend? And can people use `delay` if it's not reported?

Reply:
Good point, I moved the `if not delay_support:` early exit earlier in the function, so the warning and `'id'` search aren't performed if the device doesn't support `'delay'`.